fullrt/dht.go

```diff
@@ -0,0 +1,1326 @@
+		return nil
+	case <-dht.ctx.Done():
+		return fmt.Errorf("dht is closed")
+	}
+}
+
+func (dht *FullRT) Stat() map[string]peer.ID {
+	newMap := make(map[string]peer.ID)
+
+	dht.kMapLk.RLock()
+	for k, v := range dht.keyToPeerMap {
+		newMap[k] = v
+	}
+	dht.kMapLk.RUnlock()
+	return newMap
+}
+
+func (dht *FullRT) Ready() bool {
+	dht.rtLk.RLock()
+	lastCrawlTime := dht.lastCrawlTime
+	dht.rtLk.RUnlock()
+
+	if time.Since(lastCrawlTime) > dht.crawlerInterval {
+		return false
+	}
+
+	// TODO: This function needs to be better defined. Perhaps based on going through the peer map and seeing when the
+	// last time we were connected to any of them was.
+	dht.peerAddrsLk.RLock()
+	rtSize := len(dht.keyToPeerMap)
+	dht.peerAddrsLk.RUnlock()
+
+	return rtSize > len(dht.bootstrapPeers)+1
+}
+
+func (dht *FullRT) Host() host.Host {
+	return dht.h
+}
+
+func (dht *FullRT) runCrawler(ctx context.Context) {
+	defer dht.wg.Done()
+	t := time.NewTicker(dht.crawlerInterval)
+
+	m := make(map[peer.ID]*crawlVal)
+	mxLk := sync.Mutex{}
+
+	initialTrigger := make(chan struct{}, 1)
+	initialTrigger <- struct{}{}
+
+	for {
+		select {
+		case <-t.C:
+		case <-initialTrigger:
+		case <-dht.triggerRefresh:
+		case <-ctx.Done():
+			return
+		}
+
+		var addrs []*peer.AddrInfo
+		dht.peerAddrsLk.Lock()
+		for k := range m {
+			addrs = append(addrs, &peer.AddrInfo{ID: k}) // Addrs: v.addrs
+		}
+
+		addrs = append(addrs, dht.bootstrapPeers...)
+		dht.peerAddrsLk.Unlock()
+
+		for k := range m {
+			delete(m, k)
+		}
+
+		start := time.Now()
+		dht.crawler.Run(ctx, addrs,
+			func(p peer.ID, rtPeers []*peer.AddrInfo) {
+				conns := dht.h.Network().ConnsToPeer(p)
+				var addrs []multiaddr.Multiaddr
+				for _, conn := range conns {
+					addr := conn.RemoteMultiaddr()
+					addrs = append(addrs, addr)
+				}
+
+				if len(addrs) == 0 {
+					logger.Debugf("no connections to %v after successful query. keeping addresses from the peerstore", p)
+					addrs = dht.h.Peerstore().Addrs(p)
+				}
+
+				keep := kaddht.PublicRoutingTableFilter(dht, p)
+				if !keep {
+					return
+				}
+
+				mxLk.Lock()
+				defer mxLk.Unlock()
+				m[p] = &crawlVal{
+					addrs: addrs,
+				}
+			},
+			func(p peer.ID, err error) {})
+		dur := time.Since(start)
+		logger.Infof("crawl took %v", dur)
+
+		peerAddrs := make(map[peer.ID][]multiaddr.Multiaddr)
+		kPeerMap := make(map[string]peer.ID)
+		newRt := trie.New()
+		for k, v := range m {
+			v.key = kadkey.KbucketIDToKey(kb.ConvertPeerID(k))
+			peerAddrs[k] = v.addrs
+			kPeerMap[string(v.key)] = k
+			newRt.Add(v.key)
+		}
+
+		dht.peerAddrsLk.Lock()
+		dht.peerAddrs = peerAddrs
+		dht.peerAddrsLk.Unlock()
+
+		dht.kMapLk.Lock()
+		dht.keyToPeerMap = kPeerMap
+		dht.kMapLk.Unlock()
+
+		dht.rtLk.Lock()
+		dht.rt = newRt
+		dht.lastCrawlTime = time.Now()
+		dht.rtLk.Unlock()
+	}
+}
+
+func (dht *FullRT) Close() error {
+	dht.cancel()
+	err := dht.ProviderManager.Process().Close()
+	dht.wg.Wait()
+	return err
+}
```

Comment:
@Stebalien I added this along with changing the constructor to not take a context anymore so we can move towards better lifecycle management (instead of context cancellation shenanigans).

I think I'm using goprocess here correctly, but would appreciate a set of 👀 (looking at the relevant commit ff66a31ed04bbe8cf6be9e00516042b113d58406 is probably best)

Reply:
That looks right. But we should probably derive the context from the process so we don't need to cancel it.

Reply:
I'm going to leave this as is for now. We should be removing the use of goprocess from the ProvideManager anyhow so I'd like to avoid leaning into it.